[Driver] Add command-line flags for enforcement of law of exclusivity

devincoughlin opened the pull request:
Add an -enforce-exclusivity=... flag to control enforcement of the law of
exclusivity. The flag takes one of four options:

"checked": Perform both static (compile-time) and dynamic (run-time) checks.
"unchecked": Perform only static enforcement. This is analogous to -Ounchecked.
"dynamic-only": Perform only dynamic checks. This is for staging purposes.
"none": Perform no checks at all. This is also for staging purposes.

The default, for now, is "none".

The intent is that in the fullness of time, "checked" and "unchecked" will
be the only legal options with "checked" the default. That is, static
enforcement will always be enabled and dynamic enforcement will be enabled
by default.

devincoughlin commented:
@jrose-apple Would you mind taking a look?

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -570,4 +570,8 @@ def sanitize_coverage_EQ : CommaJoined<["-"], "sanitize-coverage=">,
 
+def enforce_exclusivity_EQ : Joined<["-"], "enforce-exclusivity=">,
+  Flags<[FrontendOption]>, MetaVarName<"<enforcement>">,
+  HelpText<"Enforce law of exclusivity.">;
```
Nitpick: no periods in HelpText.

jrose-apple reviewed (approved):
Looks good.

devincoughlin committed:
[Driver] Add command-line flags for enforcement of law of exclusivity

Add an -enforce-exclusivity=... flag to control enforcement of the law of
exclusivity. The flag takes one of four options:

"checked": Perform both static (compile-time) and dynamic (run-time) checks.
"unchecked": Perform only static enforcement. This is analogous to -Ounchecked.
"dynamic-only": Perform only dynamic checks. This is for staging purposes.
"none": Perform no checks at all. This is also for staging purposes.

The default, for now, is "none".

The intent is that in the fullness of time, "checked" and "unchecked" will
be the only legal options with "checked" the default. That is, static
enforcement will always be enabled and dynamic enforcement will be enabled
by default.

devincoughlin commented:
Thanks. Removed the extra period.

devincoughlin commented:
@swift-ci Please smoke test and merge

devincoughlin commented:
@swift-ci Please smoke test and merge
